### PR TITLE
Retry submission of pending transactions

### PR DESCRIPTION
**Pull request by rvl**
### Issue Number

ADP-764

### Overview

Adds tx resubmission to the wallet server.

   - [x] Add a wallet thread for resubmitting transactions
   - [x] Initial version resubmits every pending tx about every 10 blocks after initial submission.
   - [x] Unit test retry thread
   - [x] Design doc

### Comments

- Based on PR #2570 branch - merge that first.

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -1574,17 +1591,23 @@ submitTx ctx wid (tx, meta, binary) = db & \DBLayer{..} -> do
     withExceptT ErrSubmitTxNetwork $ traceWithExceptT tr $
         postTx nw binary
     mapExceptT atomically $ do
+        withExceptT ErrSubmitTxNoSuchWallet $
+            putLocalTxSubmission (PrimaryKey wid)
+                (tx ^. #txId) binary (meta ^. #slotNo)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:1613]: :+1:

File: lib/core/src/Cardano/Wallet.hs [hunk 2]
```diff
@@ -1621,13 +1644,105 @@ forgetTx
...
+scheduleLocalTxSubmission
+    :: Word64
+    -> SlottingParameters
+    -> LocalTxSubmissionStatus tx
+    -> SlotNo
+scheduleLocalTxSubmission numBlocks sp st =
+    (st ^. #latestSubmission) + numSlots
+  where
+    numSlots = SlotNo (ceiling (fromIntegral numBlocks / f))
+    ActiveSlotCoefficient f = getActiveSlotCoefficient sp
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:1681]: :+1:

File: lib/core/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -1621,13 +1644,105 @@ forgetTx
...
+runLocalTxSubmissionPool'
+    :: forall ctx s k m.
+        ( MonadUnliftIO m
+        , MonadMonotonicTime m
+        , HasLogger WalletLog ctx
+        , HasNetworkLayer m ctx
+        , HasDBLayer m s k ctx
+        )
+    => DiffTime -- ^ Minimum time between updates
+    -> Word64 -- ^ Retry approximately every so many blocks
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: (suggestion) Consider making those a record to make things clearer at call and consumer sites:

```hs
data LocalTxSubmissionConfig = LocalTxSubmissionConfig
    { rateLimit :: DiffTime
    , blockInterval :: Word64
    }
```

File: lib/core/src/Cardano/Wallet.hs [hunk 4]
```diff
@@ -1621,13 +1644,105 @@ forgetTx
...
+                (PrimaryKey wid)
+                (st ^. #txId)
+                (st ^. #submittedTx)
+                sl
+    watchNodeTip nw submitPending
+  where
+    nw = ctx ^. networkLayer @m
+    db = ctx ^. dbLayer @m @s @k
+
+    isScheduled sp now = (<= now) . scheduleLocalTxSubmission blockInterval sp
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: :+1:

File: lib/core/src/Cardano/Wallet.hs [hunk 5]
```diff
@@ -1621,13 +1644,105 @@ forgetTx
...
+    => DiffTime
+    -> (Time -> a -> m ())
+    -> m (a -> m ())
+throttle interval action = do
+    var <- newMVar Nothing
+    pure $ \arg -> modifyMVar_ var $ \prev -> do
+        now <- getMonotonicTime
+        if (maybe interval (diffTime now) prev >= interval)
+           then action now arg $> Just now
+           else pure prev
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:1756]: :+1:

File: specifications/design-notes/tx-resubmission.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
+# Transaction resubmission
+
+Any transaction on the Cardano blockchain can be rolled back due to a
+chain switch, until its block is _2k_ blocks deep.
```
Comment (KtorZ) [inline specifications/design-notes/tx-resubmission.md]: ~2k~ → `3k/f`  

2k was true in Ouroboros Classic / BFT but not in Praos.

File: specifications/design-notes/tx-resubmission.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+Once a submitted transaction has expired, or was accepted and can no
+longer be rolled back (i.e. it is more than _2k_ blocks deep), it is
+removed from the `local_tx_submission` table.
+
+### Exceptions
+
+Pre-signed transactions which were posted to the external transactions
+endpoint are _not_ retried -- it is up to the user to retry. This
+could be changed in future if the wallet can discover transactions
+which don't belong to its own wallets.
```
Comment (KtorZ) [inline specifications/design-notes/tx-resubmission.md:57]: :ok_hand:

File: lib/core/test/unit/Cardano/WalletSpec.hs [hunk 8]
```diff
@@ -627,6 +687,321 @@ prop_estimateFee (NonEmpty coins) =
...
+
+    monitor $ counterexample $ unlines $
+        [ "posted txs = " ++ show (resSubmittedTxs res)
+        , "final pool state = " ++ show (resAction res)
+        , "logs:"
+        ] ++ map (T.unpack . toText) (resLogs res)
+
+    -- props:
+    --  1. pending transactions in pool are retried
+    assert (all (`elem` (submittedTx <$> retryTestPool tc)) (resSubmittedTxs res))
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/WalletSpec.hs]: :+1:

File: lib/core/test/unit/Cardano/WalletSpec.hs [hunk 9]
```diff
@@ -627,6 +687,321 @@ prop_estimateFee (NonEmpty coins) =
...
+
+    --  2. non-pending transactions not retried
+    let mkSealed = SealedTx . getHash . view #txId
+    let nonPending = map (mkSealed . fst)
+            . filter ((/= Pending) . view #status . snd)
+            . getTxHistory $ retryTestTxHistory tc
+    assert (all (`notElem` (resSubmittedTxs res)) nonPending)
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/WalletSpec.hs:845]: :+1:

File: lib/core/test/unit/Cardano/WalletSpec.hs [hunk 10]
```diff
@@ -627,6 +687,321 @@ prop_estimateFee (NonEmpty coins) =
...
+    mockNetwork var = dummyNetworkLayer
+        { currentSlottingParameters = pure (testSlottingParameters tc)
+        , postTx = \tx -> ExceptT $ do
+                stash var tx
+                pure $ case lookup tx (postTxResults tc) of
+                    Just True -> Right ()
+                    Just False -> Left (W.ErrPostTxBadRequest "intended")
+                    Nothing -> Left (W.ErrPostTxProtocolFailure "unexpected")
+        , watchNodeTip = mockNodeTip (numSlots tc) 0
+        }
```
Comment (KtorZ) [inline lib/core/test/unit/Cardano/WalletSpec.hs:876]: :+1:

**Review by KtorZ — APPROVED** [review]

File: specifications/design-notes/tx-resubmission.md (hunk 6 above)
Comment (rvl) [reply specifications/design-notes/tx-resubmission.md]: Doh forgot, thanks.

File: lib/core/src/Cardano/Wallet.hs (hunk 3 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet.hs]: Good idea, fixed.

File: lib/core/test/unit/Cardano/WalletSpec.hs (hunk 8 above)
Comment (rvl) [reply lib/core/test/unit/Cardano/WalletSpec.hs]: This turned out to be a lot of effort to set up the test, for some fairly weak properties. :grimacing: 
But I'm reasonably satisfied that it works.

Comment (KtorZ) [reply lib/core/test/unit/Cardano/WalletSpec.hs]: I have to say it was also not easy to review 😅, took me some time to get the gist but I ended up being convinced by the approach.

**Comment by rvl** [comment]
Thanks for the review.
Branch is rebased now.

bors r+

**Pull request merged**